File: NeetCode/Python/Arrays/Medium/438.AllAnagrams.py
# Given two strings s and p, return an array of all the start indices of 
# p's anagrams in s. You may return the answer in any order.

def findAnagrams(s, p):
    if len(p) > len(s): return []
    
    s_dict, p_dict = {}, {}

    for i in range(len(p)):
        p_dict[p[i]] = p_dict.get(p[i], 0) + 1      
        s_dict[s[i]] = s_dict.get(s[i], 0) + 1

    result = [0] if s_dict == p_dict else []
    left = 0

    for right in range(len(p), len(s)):
        s_dict[s[right]] = s_dict.get(s[right], 0) + 1
        s_dict[s[left]] -= 1

        if s_dict[s[left]] == 0:
            s_dict.pop(s[left])

        left += 1
        if s_dict == p_dict:
            result.append(left)

    return result

print(findAnagrams("cbaebabacd", "abc"))
print(findAnagrams("abab", "ab"))

File: NeetCode/Swift/Arrays/Medium/438.AllAnagrams.swift
//Given two strings s and p, return an array of all the start indices of 
//p's anagrams in s. You may return the answer in any order.

func findAnagrams(_ s: String, _ p: String) -> [Int] {
    if p.count > s.count { return [] }

    let s = Array(s)
    var sDict = s.prefix(p.count).reduce(into: [:]) { $0[$1, default: 0] += 1 }
    var pDict = Array(p).reduce(into: [:]) { $0[$1, default: 0] += 1 }
    var result = sDict == pDict ? [0] : []

    for i in p.count..<s.count {
        var leftIndex = i - p.count
        sDict[s[i], default: 0] += 1
        sDict[s[leftIndex]]? -= 1

        if sDict[s[leftIndex]] == 0 {
            sDict.removeValue(forKey: s[leftIndex])
        }

        if sDict == pDict {
            result.append(leftIndex + 1)
        }
    }

    return result
}

print(findAnagrams("cbaebabacd", "abc"))
print(findAnagrams("abab", "ab"))